test/cpp/end2end/subchannel_metrics.cc

```diff
@@ -0,0 +1,206 @@
+
+class SubchannelMetricsTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    stats_plugin_ = grpc_core::FakeStatsPluginBuilder()
+                        .UseDisabledByDefaultMetrics(true)
+                        .BuildAndRegister();
+    kConnectionAttemptsSucceeded_ =
+        grpc_core::GlobalInstrumentsRegistryTestPeer::
+            FindUInt64CounterHandleByName(
+                "grpc.subchannel.connection_attempts_succeeded")
+                .value();
+    kOpenConnections_ = grpc_core::GlobalInstrumentsRegistryTestPeer::
+                            FindInt64UpDownCounterHandleByName(
+                                "grpc.subchannel.open_connections")
+                                .value();
+    kDisconnections_ =
+        grpc_core::GlobalInstrumentsRegistryTestPeer::
+            FindUInt64CounterHandleByName("grpc.subchannel.disconnections")
+                .value();
+    kConnectionAttemptsFailed_ =
+        grpc_core::GlobalInstrumentsRegistryTestPeer::
+            FindUInt64CounterHandleByName(
+                "grpc.subchannel.connection_attempts_failed")
+                .value();
+  }
+  void TearDown() override {
+    grpc_core::GlobalStatsPluginRegistryTestPeer::
+        ResetGlobalStatsPluginRegistry();
+  }
+  std::shared_ptr<grpc_core::FakeStatsPlugin> stats_plugin_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kConnectionAttemptsSucceeded_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kDisconnections_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kConnectionAttemptsFailed_;
+  grpc_core::GlobalInstrumentsRegistry::GlobalInstrumentHandle
+      kOpenConnections_;
+};
+
+TEST_F(SubchannelMetricsTest, SubchannelMetricsBasic) {
+  ConnectionAttemptInjector injector;
+  const int port = grpc_pick_unused_port_or_die();
+  std::string target = "127.0.0.1:" + std::to_string(port);
+  ServerBuilder builder;
+  builder.AddListeningPort("127.0.0.1:" + std::to_string(port),
+                           grpc::InsecureServerCredentials());
+  auto service = std::make_unique<MinimalEchoService>();
+  builder.RegisterService(service.get());
+  auto server = builder.BuildAndStart();
+  ASSERT_NE(server, nullptr);
+  auto channel =
+      grpc::CreateChannel(target, grpc::InsecureChannelCredentials());
+  auto stub = grpc::testing::EchoTestService::NewStub(channel);
+  auto hold = injector.AddHold(port);
+  channel->GetState(true);
+  hold->Wait();
+  hold->Resume();
+  EXPECT_TRUE(
+      WaitForChannelState(channel.get(), [](grpc_connectivity_state state) {
+        return state == GRPC_CHANNEL_READY;
+      }));
+  grpc::testing::EchoRequest request;
+  request.set_message("test");
+  grpc::testing::EchoResponse response;
+  grpc::ClientContext context;
+  context.set_deadline(std::chrono::system_clock::now() +
+                       std::chrono::seconds(1));
+  grpc::Status status = stub->Echo(&context, request, &response);
```

Comment:
I don't think we actually need to send this RPC.  Once the channel reports READY state, it will have an open connection, so the metrics should already be updated at that point.